### PR TITLE
Feature/genetic structures constuctor

### PR DESCRIPTION
**Pull request by Regis128**
Fixed OpenCvDataset error, added constructor to Individual and Generation classes, fixed libGenetic problem with Seed type and added few docs

### CODE REVIEW

File: src/Spectre.libClassifier/ObservationExtractor.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿/*
 * ObservationExtractor.h
-* 
+* class for getting data from Individual object
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/ObservationExtractor.h]: This is not necessarily true.

File: src/Spectre.libClassifier/ObservationExtractor.h [hunk 2]
```diff
@@ -18,22 +18,37 @@ limitations under the License.
...
 #include "Spectre.libDataset/Empty.h"
 #include "Spectre.libClassifier/OpenCvDataset.h"
-#include "Spectre.libGenetic/Individual.h"
 
 namespace spectre::supervised {
 
 using DataPointer = const spectre::core::dataset::IReadOnlyDataset<Observation, Label, spectre::core::dataset::Empty>*;
 
+/// <summary>
+/// Class used for getting dataset from individual.
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/ObservationExtractor.h]: Concept of `Individual` does not exist in the context of `libClassifier`.

File: src/Spectre.libClassifier/ObservationExtractor.h [hunk 3]
```diff
@@ -18,22 +18,37 @@ limitations under the License.
...
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ObservationExtractor"/> class.
+    /// </summary>
+    /// <param name="data">The data.</param>
     explicit ObservationExtractor(const DataPointer data);
-    OpenCvDataset getOpenCvDatasetFromIndividual(const spectre::algorithm::genetic::Individual &individual);
+    /// <summary>
+    /// Gets data.
+    /// </summary>
+    /// <param name="individual">The individual.</param>
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/ObservationExtractor.h]: Also here...

File: src/Spectre.libClassifier/ObservationExtractor.h [hunk 4]
```diff
@@ -18,22 +18,37 @@ limitations under the License.
...
+    /// <returns>OpenCvDataset</returns>
+    OpenCvDataset getOpenCvDatasetFromIndividual(const std::vector<bool>& individual) const;
```
Comment (gmrukwa) [inline src/Spectre.libClassifier/ObservationExtractor.h]: ... and here.

File: src/Spectre.libGenetic.Tests/GenerationTest.cpp [hunk 5]
```diff
@@ -27,6 +27,33 @@ namespace
 using namespace spectre::algorithm::genetic;
 using namespace spectre::core::exception;
 
+const auto seed = 0ul;
+
+TEST(GenerationInitialization, initializes)
```
Comment (gmrukwa) [inline src/Spectre.libGenetic.Tests/GenerationTest.cpp]: We were talking about moving this functionality (random initialization) from `Generation` to a kind of factory.

File: src/Spectre.libGenetic.Tests/IndividualTest.cpp [hunk 6]
```diff
@@ -26,13 +26,24 @@ namespace
...
+const auto seed = 0ul;
+
 TEST(IndividualInitialization, initializes)
 {
     EXPECT_NO_THROW(Individual({ true, true, true, true }));
     EXPECT_NO_THROW(Individual({ false, false, false, false }));
     EXPECT_NO_THROW(Individual({ true, false, true, false }));
 }
 
+TEST(IndividualInitialization, initializes_by_shuffle)
```
Comment (gmrukwa) [inline src/Spectre.libGenetic.Tests/IndividualTest.cpp]: Similarly, as in the case of `Generation`. We can provide single factory for random `Generation`, which creates random `Individual`s and assembles them into a `Generation`.

File: src/Spectre.libGenetic/Generation.cpp [hunk 7]
```diff
@@ -44,6 +45,19 @@ Generation::Generation(std::vector<Individual> &&generation):
     else { }
 }
 
+
+Generation::Generation(size_t size, size_t individualSize, size_t initialFillup, Seed seed)
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/Generation.cpp]: I've written above about this functionality.

File: src/Spectre.libGenetic/Individual.cpp [hunk 8]
```diff
@@ -18,17 +18,43 @@ limitations under the License.
...
 using namespace std;
+using namespace spectre::core::exception;
 
 namespace spectre::algorithm::genetic
 {
 Individual::Individual(std::vector<bool> &&binaryData):
     m_BinaryData(binaryData) { }
 
+
+Individual::Individual(size_t size, size_t initialFillup, Seed seed)
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/Individual.cpp]: ... and this one.

File: src/Spectre.libGenetic/Individual.h [hunk 9]
```diff
@@ -19,6 +19,7 @@ limitations under the License.
 
 #pragma once
 #include <vector>
+#include "DataTypes.h"
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/Individual.h]: Missing directory prefix. It won't work through NuGet package.

File: src/Spectre.libClassifier/ObservationExtractor.h (hunk 1 above)
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.h]: done

File: src/Spectre.libClassifier/ObservationExtractor.h (hunk 2 above)
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.h]: done

File: src/Spectre.libClassifier/ObservationExtractor.h (hunk 3 above)
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.h]: done

File: src/Spectre.libClassifier/ObservationExtractor.h (hunk 4 above)
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.h]: done

File: src/Spectre.libGenetic/Individual.h (hunk 9 above)
Comment (Regis128) [reply src/Spectre.libGenetic/Individual.h]: done

File: src/Spectre.libGenetic.Tests/GenerationTest.cpp (hunk 5 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/GenerationTest.cpp]: done

File: src/Spectre.libGenetic.Tests/IndividualTest.cpp (hunk 6 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/IndividualTest.cpp]: done

File: src/Spectre.libGenetic/Generation.cpp (hunk 7 above)
Comment (Regis128) [reply src/Spectre.libGenetic/Generation.cpp]: done

File: src/Spectre.libGenetic/Individual.cpp (hunk 8 above)
Comment (Regis128) [reply src/Spectre.libGenetic/Individual.cpp]: done

File: src/Spectre.libGenetic/GenerationFactory.cpp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        throw core::exception::ArgumentOutOfRangeException<size_t>("initialFillup", 0, individualSize, initialFillup);
+    }
+}
+
+Generation GenerationFactory::buildDefaultGeneration(Seed seed) const
+{
+    std::vector<Individual> individuals;
+    for (auto i = 0u; i < m_Size; i++)
+    {
+        individuals.push_back(buildDefaultIndividual(seed + i));
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/GenerationFactory.cpp]: I'd go for generating seed step basing on the initial seed.

**What's the issue?**

When we'll try to make random `Generation`s with different `seed` in loop. Most likely similarly like you're doing here - it's the simplest way: `factory.buildDefaultGeneration(0)` and then `factory.buildDefaultGeneration(1)`.

In current implementation, there is only one `Individual` different in such generations, while even for small seed difference we should get absolutely different ones!

I'd propose something like this, or similar:
```
RNG rng(seed);
const auto seedIncrement = rng.next();
...
for(...)
{
    buildIndividual(seed + i * seedIncrement);
}
...
```

I hope, you grasp the idea.

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+{
+    Generation testGeneration = generationFatoryWithBigIndividuals.buildDefaultGeneration(seed);
+    bool different = false;
+
+    EXPECT_EQ(testGeneration[0].size(), testGeneration[1].size());
+    for (auto i = 0u; i < testGeneration[0].size(); i++)
+    {
+        if (testGeneration[0][i] != testGeneration[1][i])
+        {
+            different = true;
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: done

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+    EXPECT_EQ(different, true);
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: done

File: src/Spectre.libGenetic.Tests/IndividualTest.cpp [hunk 13]
```diff
@@ -26,6 +26,8 @@ namespace
 {
 using namespace spectre::algorithm::genetic;
 
+const auto seed = 0ul;
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/IndividualTest.cpp]: forgot about it, done

File: src/Spectre.libGenetic/GenerationFactory.h [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+#include "Spectre.libGenetic/DataTypes.h"
+#include "Generation.h"
+#include <memory>
+
+namespace spectre::algorithm::genetic
+{
+/// <summary>
+/// Factory for creating Generation objects with parameters.
+/// </summary>
+class GenerationFactory
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libGenetic/GenerationFactory.h]: done

File: src/Spectre.libGenetic/GenerationFactory.h [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    /// <param name="size">Size of generation.</param>
+    /// <param name="individualSize">Size of every individual in generation.</param>
+    /// <param name="initialFillup">The initial fillup.</param>
+    /// <param name="seed">The seed.</param>
+    GenerationFactory(size_t size, size_t individualSize, size_t initialFillup);
+    /// <summary>
+    /// Creates generation with default parameter values.
+    /// </summary>
+    /// <returns>Created generation.</returns>
+    Generation buildDefaultGeneration(Seed seed) const;
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libGenetic/GenerationFactory.h]: done

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        generationFatoryWithBigIndividuals(bigGenerationSize, bigIndividualSize, bigInitialFillup) {}
+
+protected:
+    const size_t generationSize = 1u;
+    const size_t individualSize = 6u;
+    const size_t initialFillup = 4u;
+    const size_t bigGenerationSize = 2u;
+    const size_t bigIndividualSize = 100u;
+    const size_t bigInitialFillup = 60u;
+    const Seed seed = 0ul;
```
Comment (Sand3r-) [inline src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: unsigned long? :smiley:

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        EXPECT_EQ(trueAmount, initialFillup);
+    }
+}
+
+TEST_F(GenerationFactoryInitializationTest, create_different_individuals_for_different_seeds)
+{
+    Generation testGeneration = generationFatoryWithBigIndividuals(seed);
+    bool different = false;
+
+    EXPECT_EQ(testGeneration[0].size(), testGeneration[1].size());
```
Comment (Sand3r-) [inline src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: Shouldn't that be an ASSERT_EQ? If the sizes are different, we may read out of bounds of testGeneration[1].

File: src/Spectre.libClassifier/ObservationExtractor.cpp [hunk 18]
```diff
@@ -34,13 +35,13 @@ ObservationExtractor::ObservationExtractor(const DataPointer data): m_Data(data)
     }
 }
 
-OpenCvDataset ObservationExtractor::getOpenCvDatasetFromIndividual(const spectre::algorithm::genetic::Individual &individual)
+OpenCvDataset ObservationExtractor::getOpenCvDatasetFromIndividual(const std::vector<bool>& observations) const
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: done

File: src/Spectre.libClassifier/ObservationExtractor.h [hunk 19]
```diff
@@ -18,22 +18,37 @@ limitations under the License.
...
+    /// </summary>
+    /// <param name="observations">The vector of bool containing information which observations to include.</param>
+    /// <returns>OpenCvDataset</returns>
+    OpenCvDataset getOpenCvDatasetFromIndividual(const std::vector<bool>& observations) const;
 
 private:
+    /// <summary>
+    /// The dataset.
+    /// </summary>
     const DataPointer m_Data;
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.h]: good point, done

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp (hunk 17 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: another good point, done

File: src/Spectre.libClassifier/ObservationExtractor.cpp [hunk 20]
```diff
@@ -17,8 +17,9 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+#include <vector>
 #include "Spectre.libException/NullPointerException.h"
-#include "ObservationExtractor.h"
+#include "Spectre.libClassifier/ObservationExtractor.h"
 
 namespace spectre::supervised {
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: Grzegorz always does it like that, so I think we should be consistent. And he must have a reason for doing it that way, however I don't know it either

File: src/Spectre.libClassifier/ObservationExtractor.cpp [hunk 21]
```diff
@@ -34,13 +35,13 @@ ObservationExtractor::ObservationExtractor(const DataPointer data): m_Data(data)
...
 {
     std::vector<DataType> data;
     std::vector<Label> labels;
-    for (auto i = 0u; i < individual.size(); ++i)
+    for (auto i = 0u; i < observations.size(); ++i)
     {
-        if (individual[i] == true)
+        if (observations[i])
         {
             const auto& observation = m_Data->operator[](i);
```
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: I don't understand what you mean. We have to check in observations parameter if we need to include specific observation and then we add it from m_Data. Observations and m_Data contain different things

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp (hunk 16 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: Again, I dont't understand :D. Types are coherent with ones that are parameters of the functions.

File: src/Spectre.libClassifier/ObservationExtractor.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (gmrukwa) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: Yeah, I did. It was an experiment. I don't think we should do this anymore :D

File: src/Spectre.libClassifier/ObservationExtractor.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (gmrukwa) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: @Regis128: @Sand3r- means that `m_Data->operator[](i)` is too verbose if we can do `m_Data[i]`. Can we?

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp (hunk 16 above)
Comment (gmrukwa) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: @Sand3r-, this one was `unsigned long` for current RNG type indeed.

File: src/Spectre.libGenetic/Individual.cpp [hunk 22]
```diff
@@ -18,17 +18,24 @@ limitations under the License.
...
 
 using namespace std;
+using namespace spectre::core::exception;
 
 namespace spectre::algorithm::genetic
 {
 Individual::Individual(std::vector<bool> &&binaryData):
     m_BinaryData(binaryData) { }
 
+std::vector<bool> Individual::getData() const
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply src/Spectre.libGenetic/Individual.cpp]: Good catch!

File: src/Spectre.libClassifier/ObservationExtractor.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: done

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp (hunk 16 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: nothing to do here, so done

File: src/Spectre.libClassifier/ObservationExtractor.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: m_Data[i] not working

[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: tam mamy pointer, a my chcemy iść do wiersza, a nie do innej instancji w tablicy m_Data, dlatego tak jest, m_Data nie jest tablicą, także nic do zmiany

File: src/Spectre.libGenetic/Individual.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (Regis128) [reply src/Spectre.libGenetic/Individual.cpp]: done

File: src/Spectre.libClassifier/ObservationExtractor.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (Sand3r-) [reply src/Spectre.libClassifier/ObservationExtractor.cpp]: Alrighty then

File: src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp (hunk 16 above)
Comment (Sand3r-) [reply src/Spectre.libGenetic.Tests/GenerationFactoryTest.cpp]: @gmrukwa Funny how other seeds are just unsigned ints then xd.

File: src/Spectre.libGenetic/GenerationFactory.h [hunk 23]
```diff
@@ -0,0 +1,66 @@
+﻿/*
+* GenerationFactory.h
+* Builds different versions of Generation.
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/GenerationFactory.h]: Please update the description to match the docstring.

**Comment by gmrukwa** [comment]
[Merge updates of GA](https://trello.com/c/rHY973Dc/344-merge-updates-of-ga)

**Pull request merged**